File: client/android/app/src/main/java/app/mobiling/client/cart/CartFeatureBridge.kt
package app.mobiling.client.cart

import app.mobiling.client.contract.cart.AddCartItemRequest
import app.mobiling.client.contract.cart.MobileCartCheckoutHandoffPayload
import app.mobiling.client.contract.cart.MobileCartPayload
import app.mobiling.client.data.cart.CartCheckoutGateway
import app.mobiling.client.data.cart.CartReader
import app.mobiling.client.data.cart.CartWriter
import app.mobiling.client.usecase.cart.AddCartItemUseCase
import app.mobiling.client.usecase.cart.LoadCurrentCartUseCase
import app.mobiling.client.usecase.cart.PrepareCartCheckoutHandoffUseCase

/**
 * Marketing America Corp. Oleksandr Tishchenko
 */
class CartFeatureBridge(
    private val reader: CartReader,
    private val writer: CartWriter,
    private val checkoutGateway: CartCheckoutGateway,
) {
    suspend fun current(): MobileCartPayload =
        LoadCurrentCartUseCase(reader).invoke()

    suspend fun add(request: AddCartItemRequest): MobileCartPayload =
        AddCartItemUseCase(writer).invoke(request)

    suspend fun checkoutHandoff(): MobileCartCheckoutHandoffPayload =
        PrepareCartCheckoutHandoffUseCase(checkoutGateway).invoke()

}

File: client/ios/App/Cart/CartFeatureBridge.swift
import Foundation

public struct CartFeatureBridge: Sendable {
    private let reader: any CartReader
    private let writer: any CartWriter
    private let checkoutGateway: any CartCheckoutGateway

    public init(reader: any CartReader, writer: any CartWriter, checkoutGateway: any CartCheckoutGateway) {
        self.reader = reader
        self.writer = writer
        self.checkoutGateway = checkoutGateway
    }

    public func current() async throws -> CartMobilePayload {
        try await CartLoadCurrentUseCase(reader: reader)()
    }

    public func add(request: CartAddItemRequest) async throws -> CartMobilePayload {
        try await CartAddItemUseCase(writer: writer)(request: request)
    }

    public func checkoutHandoff() async throws -> CartCheckoutHandoffPayload {
        try await CartPrepareCheckoutHandoffUseCase(gateway: checkoutGateway)()
    }
}
